Pass a payload with useful information into Java validators

mkurz opened the pull request:
Right now, inside a validator's `isValid` or a form's `validate()` method there is no way to access context information which may be needed for a validation process.
The only workaround is to make use of the ThreadLocal `Http.Context.current` and e.g. put information into it's request `args` or get the current request's language out of it:

```java
@Override
public boolean isValid(final Object object) {
    // The only way to access contextual information for the current validation process:
    // Http.Context.current().args.get("somekey") or
    // Http.Context.current().lang() or
    // Http.Context.current().messages()
    // etc.
}

// The same problem applies to the validate() methods inside Form classes:
@Override
public ValidationError validate() {
    // How to access contextual information in here?
    // E.g. we want to access the current user saved in the ctx.args or
    // need the lang to validate numbers, dates,...
}
```

Because `Http.Context.current` will likely be removed in future (and also the above code is just not nice anyway even if we don't remove it...) I worked closely with the Hibernate Validator team (thanks @gsmet, @gunnarmorling and @marko-bekhta) to find a solution for this problem (see [HV-1529](https://hibernate.atlassian.net/browse/HV-1529) / https://github.com/hibernate/hibernate-validator/pull/890, [HV-1595](https://hibernate.atlassian.net/browse/HV-1595) / https://github.com/hibernate/hibernate-validator/pull/942, [HV-1589](https://hibernate.atlassian.net/browse/HV-1589) / https://github.com/hibernate/hibernate-validator/pull/939):
So with [today's](http://in.relation.to/2018/03/27/hibernate-validator-609-final-out/) Hibernate Validator 6.0.9 release we are now able to pass a payload into the validators' `isValid` methods!

For Play that means a `ValidatorPayload` object with useful information (sometimes) needed for a validation process gets passed to an `isValid(...)` and `validate(...)` method now:
```java
@Override
public boolean isValid(final Object object, ValidatorPayload payload {
    payload.getLang(); // language of the current request
    payload.getMessages(); // messages defined for the current request lang
    payload.getArgs(); // request context args
    // ...
}

@Override
public String validate(ValidatorPayload payload) {
    // same same but different
}
```

I used a new `ValidatorPayload` class so it's typesafe (instead of just passing a map) and we can also add fields to it later if we want to pass more stuff.

For me this is the last missing piece in the Java form's validation puzzle :smile:

mkurz commented:
I will update this PR next week so it doesn't break compatibility - by making the payload an optional feature.

mkurz commented:
I updated this pr. Now it doesn't break compatibility but makes the payload an optional feature.

Waiting for review :smile:

PromanSEW commented on framework/src/play-java-forms/src/main/java/play/data/validation/Constraints.java:
```diff
@@ -644,6 +704,81 @@ public boolean isValid(Object object) {
+    @Retention(RUNTIME)
+    @Constraint(validatedBy = ValidatePayloadWithValidator.class)
+    @Repeatable(play.data.validation.Constraints.ValidatePayloadWith.List.class)
+    @play.data.Form.Display(name="constraint.validatewith", attributes={})
```
This class is already imported. You can replace it by `@Display` here and in other places

mkurz replied:
I updated the pr. Thanks!

marcospereira commented on documentation/manual/working/javaGuide/main/forms/code/javaguide/forms/customconstraint/ValidatableWithDB.java:
```diff
@@ -6,7 +6,13 @@
 //#interface
 import play.db.Database;
 
+//import play.data.validation.Constraints.ValidatorPayload;
```
Remove comment?

mkurz replied:
Removed!

marcospereira commented on documentation/manual/working/javaGuide/main/forms/code/javaguide/forms/customconstraint/ValidateWithDBValidator.java:
```diff
@@ -28,5 +31,11 @@ public void initialize(final ValidateWithDB constraintAnnotation) {
         return reportValidationStatus(value.validate(this.db), constraintValidatorContext);
     }
+
+    // or, if you want to pass a payload:
```
I would be better to have distinct non-commented snippets instead. It will be harder to keep these updated since they won't fail compilation if we change the API later.

mkurz replied:
I now added distinct code snippets for examples that use a validation payload and for examples don't. In the docs I used the tab style like we do for Java vs. Scala code snippets - but named the tabs "Without payload" vs "With payload".

marcospereira commented on framework/src/play-java-forms/src/main/java/play/data/DynamicForm.java:
```diff
@@ -32,10 +32,10 @@
      */
-    public DynamicForm(MessagesApi messagesApi, Formatters formatters, Validator validator) {
-        super(DynamicForm.Dynamic.class, messagesApi, formatters, validator);
+    public DynamicForm(MessagesApi messagesApi, Formatters formatters, ValidatorFactory validatorFactory) {
```
I think we may have some MiMa configuration issue since these are not reporting any problem when changing the parameter type here.

marcospereira replied:
Hmm...

https://github.com/playframework/playframework/blob/master/framework/project/BuildSettings.scala#L346

Looks like this is filtering any change to the constructor?

marcospereira replied:
The same is happening for `Form` too.

mkurz replied:
Yeah, looks like the lines you referenced are filtering any change to the constructors.
However, this shouldn't really matter for framework users. A Play framework user is usually not effected by this breaking changes because the way to get a form is `Form<User> userForm = formFactory.form(User.class);` and not by using the constructors `new Form(...)` and `new DynamicForm(...)`.

marcospereira commented on framework/src/play-java-forms/src/main/java/play/data/Form.java:
```diff
@@ -160,7 +164,7 @@ public Form(String rootName, Class<T> clazz, Map<String,String> data, List<Valid
         this.messagesApi = messagesApi;
         this.formatters = formatters;
-        this.validator = validator;
+        this.validatorFactory = validatorFactory;
```
When inspecting the MiMa filters I saw that `ValidatorFactory` implements `java.lang.AutoCloseable` so we may need to add a shutdown hook here to close it.

marcospereira commented on framework/src/play-java-forms/src/main/java/play/data/Form.java:
```diff
@@ -347,6 +362,10 @@ private DataBinder dataBinder(String... allowedFields) {
             dataBinder.bind(new MutablePropertyValues(objectData));
+            final Http.Context ctx = Http.Context.current.get();
+            // We always pass a payload, however if there is no current http request the request specific lang, messages, args,.. will not be set
+            final ValidatorPayload payload = (ctx != null) ? new ValidatorPayload(ctx.lang(), ctx.messages(), ctx.args) : new ValidatorPayload(null, null, null);
```
`new ValidatorPayload(null, null, null)` -> `ValidatorPayload.empty()`?

mkurz replied:
Implemented!

marcospereira commented on framework/project/BuildSettings.scala:
```diff
@@ -461,7 +461,12 @@ object BuildSettings {
+      // Add ValidatorPayload to Java isValid/validate methods
+      ProblemFilters.exclude[IncompatibleMethTypeProblem]("play.data.FormFactory.this"),
+      ProblemFilters.exclude[DirectMissingMethodProblem]("play.data.validation.ValidatorsComponents.validator"),
+      ProblemFilters.exclude[MissingClassProblem]("play.data.validation.ValidatorProvider")
```
Is it possible to deprecate this for later removal?

mkurz replied:
Framework users shouldn't be effected by breaking the `FormFactory` constructor because they don't use `new FormFactory(...)` but, again, the way to get a form is `Form<User> userForm = formFactory.form(User.class);`.

If you want we can keep the `ValidatorsComponents.validator` method and the `ValidatorProvider` class and deprecated all of this, however I don't know if that makes sense. I don't think we will break apps by removing it, because it's just needed by Play DI setup.

mkurz commented on documentation/manual/releases/release27/Highlights27.md:
```diff
@@ -49,6 +49,51 @@ public class MyForm {
+import play.data.validation.Constraints.ValidateWithPayload;
+import play.data.validation.Constraints.ValidatorPayload;
+
+@ValidateWithPayload
```
No, that is not possible, you need both: The annotation is needed to define a "class-level constraint" (see [HV docs](https://docs.jboss.org/hibernate/validator/5.4/reference/en-US/html_single/#section-class-level-constraints)) and the interface is needed so the form can build a contract with class-level constraint.
The class-level constraint basically executes the validation, however it needs a `validate` method - which in our case is provided via a interface that both the class-level constraint and the form know about.

mkurz commented on documentation/manual/releases/release27/Highlights27.md:
```diff
@@ -49,6 +49,51 @@ public class MyForm {
+    //...
+
+    @Override
+    public boolean isValid(final SomeValidatableInterface<?> value, final ValidatorPayload payload, final ConstraintValidatorContext constraintValidatorContext) {
```
I added a note to explain the difference.

mkurz commented on documentation/manual/working/javaGuide/main/forms/JavaForms.md:
```diff
@@ -193,6 +193,13 @@ Using this group sequence will first validate all fields belonging to the `Defau
 
+### Passing payloads to validators
+
+If needed, you can also pass a `ValidatorPayload` object - which contains useful information sometimes needed for a validation process - to a `validate` method.
```
I used `ValidationPayload` to not collide with Hibernate Validator's [`ValidationContext`](https://github.com/hibernate/hibernate-validator/blob/6.0.9.Final/engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java) (which is actually defined by the Bean Validation specs).

mkurz commented on documentation/manual/working/javaGuide/main/forms/code/javaguide/forms/JavaForms.java:
```diff
@@ -527,4 +530,27 @@ public Result index() {
+    //###insert: import play.data.validation.ValidatorPayload;
+
+    @ValidateWithPayload
+    public static class ChangePasswordForm implements ValidatableWithPayload<ValidationError> {
```
Done.

mkurz commented on documentation/manual/working/javaGuide/main/forms/code/javaguide/forms/JavaForms.java:
```diff
@@ -527,4 +530,27 @@ public Result index() {
+
+        @Override
+        public ValidationError validate(ValidatorPayload payload) {
+            payload.getLang(); // language of the current request
```
Changed.

mkurz commented on documentation/manual/working/javaGuide/main/forms/code/javaguide/forms/customconstraint/DBAccessForm.java:
```diff
@@ -6,6 +6,7 @@
 //#user
 import play.data.validation.Constraints;
 import play.data.validation.ValidationError;
+//import play.data.validation.Constraints.ValidatorPayload;
```
Removed!

mkurz commented:
@marcospereira Thanks for you feedback! I addressed all of your comments. I think everything is done. I also think the MiMa stuff should be ok, because even when breaking the `Form` and `DynamicForm` class constructors a user uses `formFactory.form(MyForm.class)` anyway to get a form instance - that is also the documented way of how to do things.

Please have a look again, thank you!

mkurz commented:
OK, to be fair, a user may uses `new DynamicForm(...)` or `new Form(...)` instead of `formFactory.form(...)` when writing tests. I will add a note to the migration guide that the constructors changed for such cases.

mkurz commented:
Alright, I think that's it. I added migration notes for the constructor changes and also kept the `ValidatorProvider` class and the `ValidatorsComponents.validator` method for now but deprecated it.
@marcospereira Review welcome, thanks!

mkurz commented:
@marcospereira This one should be ready to merge as well now :wink:

mkurz commented on documentation/manual/releases/release27/Highlights27.md:
````diff
@@ -49,6 +49,56 @@ public class MyForm {
+}
+```
+
+> Don't get confused with `ValidationPayload` and `ConstraintValidatorContext`: The former class is provided by Play and is what you use in your day-to-day work when dealing with forms in Play. The latter class is defined by the Bean Validation specification and is used only internally in Play - with one exception: This class emerges when your write your own custom class-level constraints, like i ...
````
Done.

mkurz commented on documentation/manual/working/javaGuide/main/forms/JavaForms.md:
```diff
@@ -203,19 +210,29 @@ Because constraints support both [[runtime Dependency Injection|JavaDependencyIn
 As you can see we inject the `Database` object into the constraint's constructor and use it later when calling `validate`. When using runtime Dependency Injection, Guice will automatically inject the `Database` object, but for compile-time Dependency Injection you need to do that by yourself:
 
-@[constraint-compile-timed-di](code/javaguide/forms/customconstraint/ValidateWithDBComponents.java)
+@[constraint-compile-timed-di](code/javaguide/forms/customconstraint/nopayload/ValidateWithDBComponents.java)
```
No. Actually this example isn't about payload or nopayload, it's about how to inject a `Database` object when using ompile-time DI. I moved the example out of the `nopayload` package to clarify that.

mkurz commented on framework/src/play-java-forms/src/main/java/play/data/Form.java:
```diff
@@ -325,7 +340,7 @@ private DataBinder dataBinder(String... allowedFields) {
             dataBinder.setAllowedFields(allowedFields);
         }
-        SpringValidatorAdapter validator = new SpringValidatorAdapter(this.validator);
+        SpringValidatorAdapter validator = new SpringValidatorAdapter(this.validatorFactory.getValidator());
```
No need to memoize it. `getValidator` is a very lightweight call - actually HV maintains *a lot* of internal caches, so creating a new (default) validator is basically just object instantiation. I can confirm that since I was working heavily with the HV codebase when adding the needed feature for this patch.
Also [this statement](https://github.com/hibernate/hibernate-validator/pull/890#discussion_r153158631) from gsmet (HV maintainer) backs my thought:
>  ...The drawback is that before your calls to `validate()` you would have to instantiate the `Validator` first by unwrapping the context and setting the payload. **The `Validator` creation is quite lightweight so I don't think it will be an issue in practice**....

mkurz commented:
@marcospereira Thanks. See the last 4 commits. I addressed your feedback, upgraded HV once more, updated also the form doc page to Bean Validation 2.0/HV 6 and moved a doc class out of a "wrong" package.

Should be ready to merge.

mkurz commented:
Once more resolved a conflict... Waiting for travis...

mkurz commented:
@marcospereira The sbt-plugins-0.13 test fails, but completely unrelated to this pull request. I restarted it like 5 times without success..

mkurz commented:
Green! I had to delete **all** repository caches...

marcospereira commented:
> Green! I had to delete all repository caches...

Yes, I wonder if the build will be more stable without cache (although slower) or if we need to change our cache configuration somehow. I've made this for Lagom sometime ago in https://github.com/lagom/lagom/pull/1523, but we still have some cache problems there too.
